crates/ruff_linter/src/rules/pylint/rules/invalid_string_characters.rs

```diff
@@ -201,8 +219,97 @@ pub(crate) fn invalid_string_characters(
         let location = range.start() + TextSize::try_from(column).unwrap();
         let range = TextRange::at(location, c.text_len());
 
-        diagnostics.push(Diagnostic::new(rule, range).with_fix(Fix::safe_edit(
-            Edit::range_replacement(replacement.to_string(), range),
-        )));
+        new_diagnostics.push(InvalidCharacterDiagnostic {
+            diagnostic: Diagnostic::new(rule, range),
+            // This is integrated with other fixes and attached to the diagnostic below.
+            edit: Edit::range_replacement(replacement.to_string(), range),
+        });
+    }
+    if new_diagnostics.is_empty() {
+        // No issues, nothing to fix.
+        return;
+    }
+
+    // Convert raw strings to non-raw strings when fixes are applied:
+    // https://github.com/astral-sh/ruff/issues/13294#issuecomment-2341955180
+    let mut string_conversion_edits = Vec::new();
+    if token.is_raw_string() {
+        let string_flags = token.string_flags();
+        let prefix = string_flags.prefix().as_str();
+
+        // 1. Remove the raw string prefix.
+        for (column, match_) in prefix.match_indices(&['r', 'R']) {
+            let c = match_.chars().next().unwrap();
+
+            let entire_string_range = match kind {
+                TokenKind::String => range,
+                _ => last_fstring_start.unwrap().range(),
+            };
+            let location = entire_string_range.start() + TextSize::try_from(column).unwrap();
+            let range = TextRange::at(location, c.text_len());
+
+            string_conversion_edits.push(Edit::range_deletion(range));
+        }
+
+        // 2. Escape '\' and quote characters inside the string content.
+        let (content_start, content_end): (TextSize, TextSize) = match kind {
+            TokenKind::String => (
+                prefix.text_len() + string_flags.quote_len(),
+                TextSize::try_from(text.len()).unwrap() - string_flags.quote_len(),
+            ),
+            _ => (0.into(), text.len().try_into().unwrap()),
+        };
```

Comment:
You can use `text_len` and return a `TextRange`. Returning a text range has the advantage that you can index directly by range

let string_content = &text[range];

```suggestion
                text.text_len() - string_flags.quote_len(),
            ),
            _ => TextRange::new((0.into(), text.text_len()),
        };
```